lib/voicemail-admin.js

```diff
@@ -0,0 +1,779 @@
+  } else {
+    commandName = entryComponents[1] + ' ' + entryComponents[2];
+  }
+
+  chosenCommand = menu.commands[commandName];
+
+  if (!chosenCommand) {
+    throw new Error('Command: \'%s\' does not exist.');
+  }
+
+  logger.info('Usage: ' + chosenCommand.usage.green);
+  logger.info(chosenCommand.description);
+
+  chosenCommand.details.forEach(function (detail) {
+    logger.info(detail);
+  });
+}
+
+/**
+ * Removes keys from an associative array that are either functions or objects
+ *
+ * @param {object} associativeArray - associative array to cull junk from
+ */
+function removeUgly(associativeArray) {
+  for (var key in associativeArray) {
+    var fieldType = typeof(associativeArray[key]);
+    if (fieldType === 'function' ||
+        fieldType === 'object') {
+      delete associativeArray[key];
+    }
+  }
+}
+
+/**
+ * Throws an invalid syntax error
+ *
+ * @param {array} entryComponents - full command as it was entered
+ */
+function invalidSyntax(entryComponents) {
+  throw new Error(sprintf('Invalid Syntax for \'%s %s\'',
+    entryComponents[0], entryComponents[1]));
+}
+
+/**
+ * Returns true, prints a cute little exit message
+ */
+function exit() {
+  logger.info('So long and thanks for all the fish'.cyan);
+  return true;
+}
+
+/**
+ * Returns full suite of commands needed to defined in menu.json
+ *
+ * All commands have arguments of chosenCommand and entryComponents
+ * chosenCommand is the associative array that defines details about
+ * the command currently in execution. entryComponents is an array
+ * containing the user-inputted elements of the command on a word
+ * by word basis (though words can be joined into a single element by
+ * putting them between quotation marks).
+ */
+function createApi() {
+  return {
+    showHelp: function(chosenCommand, entryComponents) {
+      if (entryComponents.length >= 2) {
+        return showHelpCommand(entryComponents);
+      }
+
+      logger.info(sprintf('%-70s'.underline, 'Command List:'));
+      for (var key in menu.commands) {
+        var command = menu.commands[key];
+        logger.info(sprintf('%s'.green + ' - %s',
+                            key, command.description));
+      }
+    },
+
+    exit: function(chosenCommand, entryComponents) {
+      return exit();
+    },
+
+    showContexts: function(chosenCommand, entryComponents) {
+      return dal.context.all()
+        .then(function(contexts) {
+          logger.info('domain          '.underline);
+          contexts.forEach(function (context) {
+            logger.info(context.domain);
+          });
+        });
+    },
+
+    showFolders: function(chosenCommand, entryComponents) {
+      return dal.folder.all()
+        .then(function(folders) {
+
+          logger.info(sprintf('%-12s %-6s %-30s'.underline,
+                              'name', 'dtmf', 'recording'));
+          for (var key in folders) {
+            var folder = folders[key];
+
+            if (key === 'add') {
```

Comment:
Since you know what the folders collection object looks like, just check to see if the type of the key is a function instead, this is a bit hard to figure out without knowing what folders looks like. Also comment as to why this is necessary.